### PR TITLE
Gabe/firebase-auth

### PR DESCRIPTION
**Pull request by gabriel-jt**
Added a basic login screen, implemented user auth, created protected routes, and verified that a logged-in user is an admin.

Updates to latest commits:
- Used Server Side Rendering to load admin data before client loads pages (prevents non-admin users from getting previews of protected routes and allows pages to properly load)
- Fixed the refreshing issue! (no longer logged out after one refresh)
- Refactored code and simplified our auth firebase lib/utility functions
- Used cookies/tokens to keep track if user has already signed in

## Testing Done (Manual)
- Navigated to every protected route (users, transactions, admins, profile) with a valid admin user and made sure I could access those pages.
- Refreshed at every protected route while logged in as an admin and made sure admin was not logged out
- Attempted to log in with non-admin credentials (ordinary user) and made sure error messages appeared on the console
- Attempted to navigate to `localhost:3000/{route}` where route is every protected route while logged out and made sure I was redirected back to the sign in page.
- Checked that unsuccessful attempts to log in cleared the sign in form

## Affected PRs
Any PR that needs admin credentials/information will now be able to!

## Screenshots

Screen Shot 2021-10-28 at 2 31 05 AM

CC: @cindyxzhang

### CODE REVIEW

**Comment by jacobjk01** [comment]
Refreshing the page signs me out! plz check and see if that's what happens for u too

File: firebase/auth/auth.tsx [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+const auth = firebaseApp.auth();
+
+export type AuthData = {
+    authUser: Admin,
+    loading: Boolean,
+}
+
+// TODO handle/throw errors accordingly
+export const signInWithEmailAndPassword = async (email: string, password: string) => {
```
Comment (sydbui) [inline firebase/auth/auth.tsx:14]: add return type for typescript

File: firebase/auth/auth.tsx [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    try {
+        await auth.signInWithEmailAndPassword(email, password);
+    } catch (e) {
+        console.error(e);
+        throw e;
+    }
+}
+
+export const registerWithEmailAndPassword = async (email: string, password: string) => {
```
Comment (sydbui) [inline firebase/auth/auth.tsx:23]: return type

File: firebase/auth/auth.tsx [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    try {
+        await auth.createUserWithEmailAndPassword(email, password);
+    } catch (e) {
+        console.error(e);
+        throw e;
+    }
+    // TODO make sure an entry with corresponding info is created in the admins table
+}
+
+const useFirebaseAuth = () => {
```
Comment (sydbui) [inline firebase/auth/auth.tsx:33]: return type

File: firebase/auth/useAuth.tsx [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+ */
+const AuthUserProvider: React.FunctionComponent = ({ children }) => {
+    const auth = useFirebaseAuth();
+    return <authUserContext.Provider value={auth}>{children}</authUserContext.Provider>
+}
+
+/**
+ * useAuth hook to get data about signed in user
+ */
+export const useAuth = () => useContext(authUserContext);
```
Comment (sydbui) [inline firebase/auth/useAuth.tsx:20]: return type

File: firebase/auth/useAuth.tsx [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+const authUserContext = createContext<AuthData>({
+    authUser: null,
+    loading: true,
+});
+
+/**
+ * Use context to provide which user is signed in to all pages
+ */
+const AuthUserProvider: React.FunctionComponent = ({ children }) => {
```
Comment (sydbui) [inline firebase/auth/useAuth.tsx:12]: props validation

**Review by sydbui — CHANGES_REQUESTED** [review]
It signs out when you refresh or even if it is a browser refresh which is pretty buggy. Also idk if this is from the auth but the transaction page looks like this for me
Screen Shot 2021-10-29 at 2 57 20 PM

**Review by cindyxzhang — CHANGES_REQUESTED** [review]
Not sure why but I get logged out when I press on the Sunday Friends text on the nav bar - I don't think this is expected behaviour for when I am logged in.

**Comment by gabriel-jt** [comment]
> Not sure why but I get logged out when I press on the Sunday Friends text on the nav bar - I don't think this is expected behaviour for when I am logged in.

I don't think you're actually logged out, I believe clicking the Sunday Friends text redirects you back to the sign in page, but you're still logged in. I believe Danashi will fix this in his nav bar PR? (making it so that the logo/sunday friends text will not navigate you back to the sign in page).

**Comment by cindyxzhang** [comment]
> It signs out when you refresh or even if it is a browser refresh which is pretty buggy. Also idk if this is from the auth but the transaction page looks like this for me Screen Shot 2021-10-29 at 2 57 20 PM

also still getting this bug on most recent commit. I'm also getting these errors in console (I know it's not the most recent version of Jacob's work) but could it be because of this? or is this auth related @jacobjk01 
image

**Comment by gabriel-jt** [comment]
@cindyxzhang I will add checks to redirect people who've already signed in back to transactions if they go back to the landing page

**Comment by gabriel-jt** [comment]
> also still getting this bug on most recent commit. I'm also getting these errors in console (I know it's not the most recent version of Jacob's work) but could it be because of this? or is this auth related @jacobjk01

@cindyxzhang I think it potentially might be coming from the transactions table. I believe some keys are not unique since a transaction item has the username as its key (which is not gonna be unique to every transaction) @jacobjk01 could you look into this for your current PR?

**Comment by gabriel-jt** [comment]
> > It signs out when you refresh or even if it is a browser refresh which is pretty buggy. Also idk if this is from the auth but the transaction page looks like this for me Screen Shot 2021-10-29 at 2 57 20 PM
> 
> also still getting this bug on most recent commit. I'm also getting these errors in console (I know it's not the most recent version of Jacob's work) but could it be because of this? or is this auth related @jacobjk01 image

Hmmm I don't think this is coming from auth... the other pages load fine for me and I no longer see this bug on my end. What exactly did you do? I wanna see if I can reproduce this bug

**Review by cindyxzhang — APPROVED** [review]
LGTM

**Pull request merged**